src/dscanner/analysis/vcall_in_ctor.d

```diff
@@ -6,433 +6,199 @@ module dscanner.analysis.vcall_in_ctor;
-            {
-                foo();
-                bar(); /+
-                ^^^ [warn]: %s +/
-            }
-            private: public void bar();
-            public private {void foo(){}}
-        }
-    }c.format(VcallCtorChecker.MSG), sac);
-
-    // passes
-    assertAnalyzerWarnings(q{
-        class Bar
-        {
-            this(){foo();}
-            private void foo(){}
-        }
-    }, sac);
-
-    assertAnalyzerWarnings(q{
-        class Bar
-        {
-            this(){foo();}
-            private {void foo(){}}
-        }
-    }, sac);
-
-    assertAnalyzerWarnings(q{
-        class Bar
-        {
-            this(){foo();}
-            private public protected private void foo(){}
-        }
-    }, sac);
-
-    assertAnalyzerWarnings(q{
-        class Bar
-        {
-            this(){foo();}
-            final private public protected void foo(){}
-        }
-    }, sac);
-
-    assertAnalyzerWarnings(q{
-        class Bar
-        {
-            this(){foo();}
-            final void foo(){}
-        }
-    }, sac);
-
-    assertAnalyzerWarnings(q{
-        final class Bar
-        {
-            public:
-            this(){foo();}
-            void foo(){}
-        }
-    }, sac);
-
-    assertAnalyzerWarnings(q{
-        class Bar
-        {
-            public:
-            this(){foo();}
-            void foo(T)(){}
-        }
-    }, sac);
-
-    assertAnalyzerWarnings(q{
-        class Foo
-        {
-            static void nonVirtual();
-            this(){nonVirtual();}
-        }
-    }, sac);
-
-    assertAnalyzerWarnings(q{
-        class Foo
-        {
-            package void nonVirtual();
-            this(){nonVirtual();}
-        }
-    }, sac);
-
-    assertAnalyzerWarnings(q{
-        class C {
-            static struct S {
-            public:
-                this(int) {
-                    foo();
-                }
-                void foo() {}
-            }
-        }
-    }, sac);
-
-    import std.stdio: writeln;
-    writeln("Unittest for VcallCtorChecker passed");
+	import dscanner.analysis.config : StaticAnalysisConfig, Check, disabledConfig;
+	import dscanner.analysis.helpers : assertAnalyzerWarningsDMD;
+	import std.stdio : stderr;
+	import std.format : format;
+
+	StaticAnalysisConfig sac = disabledConfig();
+	sac.vcall_in_ctor = Check.enabled;
+
+	enum MSG = "a virtual call inside a constructor may lead to unexpected results in the derived classes";
+
+	// fails
+	assertAnalyzerWarningsDMD(q{
+		class Bar
+		{
+			this() { foo(); } // [warn]: %s
+			private:
+			public
+			void foo() {}
+		}
+	}c.format(MSG), sac, true);
+
+	assertAnalyzerWarningsDMD(q{
+		class Bar
+		{
+			this()
+			{
+				foo(); // [warn]: %s
+				foo(); // [warn]: %s
+				bar();
+			}
+			private: void bar();
+			public { void foo() {} }
+		}
+	}c.format(MSG, MSG), sac, true);
+
+	assertAnalyzerWarningsDMD(q{
+		class Foo
+		{
+			this()
+			{
+				class Bar
+				{
+					this() { bar(); }
```

Comment:
Try calling `foo` here, I suspect that a warning will be issued although I don't think it should. To make sure, you should validate the behavior with what dscanner did before.